src/index.js

```diff
@@ -607,6 +607,95 @@ export default function (babel) {
     path.unshiftContainer("body", declarations);
   }
 
+  function generateForInIterator (path, type: "array" | "object") {
+    let idx = path.node.idx || path.scope.generateUidIdentifier("i");
+    let keys = path.scope.generateUidIdentifier("keys");
+    let len = path.scope.generateUidIdentifier("len");
+
+    let initDeclarations = [
+      t.variableDeclarator(idx, t.numericLiteral(0))
+    ]
+
+    let refId = path.node[type];
+
+    // if the target of iteration is a complex expression,
+    // hoist it into an upper declaration
+    if (!path.get(type).isIdentifier()) {
+      // _arr or _obj
+      refId = path.scope.generateUidIdentifier(type.slice(0, 3));
+      initDeclarations.unshift(
```

Comment:
I guess in terms of the usual loop, yeah, but generally you'd have the object you're iterating over first ( ie. declaring it before the loop, like it was hoisted ) so I threw it in first position. Can move it if you want.

It has to come before we use `Object.keys()` though, so the logic would be complicated by rearranging stuff.